crates/sui-core/src/consensus_adapter.rs

```diff
@@ -921,37 +932,21 @@ impl ConsensusAdapter {
                 transactions[0].kind,
                 ConsensusTransactionKind::UserTransaction(_)
             );
-        let send_end_of_publish = if is_user_tx {
-            // If we are in RejectUserCerts state and we just drained the list we need to
-            // send EndOfPublish to signal other validators that we are not submitting more certificates to the epoch.
-            // Note that there could be a race condition here where we enter this check in RejectAllCerts state.
-            // In that case we don't need to send EndOfPublish because condition to enter
-            // RejectAllCerts is when 2f+1 other validators already sequenced their EndOfPublish message.
-            // Also note that we could sent multiple EndOfPublish due to that multiple tasks can enter here with
-            // pending_count == 0. This doesn't affect correctness.
-            if epoch_store
-                .get_reconfig_state_read_lock_guard()
-                .is_reject_user_certs()
-            {
-                let pending_count = epoch_store.pending_consensus_certificates_count();
-                debug!(epoch=?epoch_store.epoch(), ?pending_count, "Deciding whether to send EndOfPublish");
-                pending_count == 0 // send end of epoch if empty
-            } else {
-                false
-            }
-        } else {
-            false
-        };
-        if send_end_of_publish {
+        if is_user_tx
+            && epoch_store.should_send_end_of_publish()
```

Comment:
nit: Put the check for has_sent_own_end_of_publish inside of should_send_end_of_publish